apps/emqx_auth/src/emqx_authn/emqx_authn_chains.erl

```diff
@@ -802,24 +835,55 @@ hook() ->
 unhook() ->
     ok = emqx_hooks:del('client.authenticate', {?MODULE, authenticate, []}).
 
-do_create_authenticator(AuthenticatorID, #{enable := Enable} = Config, Providers) ->
+do_create_authenticator(AuthenticatorID, Config, Providers) ->
     Type = authn_type(Config),
     case maps:get(Type, Providers, undefined) of
         undefined ->
-            {error, {no_available_provider_for, Type}};
+            {error, #{cause => "no_available_provider", type => Type}};
         Provider ->
-            case Provider:create(AuthenticatorID, Config) of
-                {ok, State} ->
-                    Authenticator = #authenticator{
-                        id = AuthenticatorID,
-                        provider = Provider,
-                        enable = Enable,
-                        state = State
-                    },
-                    {ok, Authenticator};
-                {error, Reason} ->
-                    {error, Reason}
-            end
+            do_create_authenticator2(AuthenticatorID, Config, Provider)
+    end.
```

Comment:
Nit:
maybe a good case for maybe, smth like:
```erlang
maybe
  Type = authn_type(Config),
  {ok, Provider} ?= find_provider(Type, Providers),
  {ok, Precondition} ?= compile_precondition(Config),
  {ok, State} ?= Provider:create(AuthenticatorID, Config),
  #{enable := Enable} = Config,
  {ok, #authenticator{...}}
end
```

Reply:
A second thought, maybe won't fix this nit.
Current style incites better code coverage.